### ISSUE TITLE
Clone time of METADATA.jl

### ISSUE THREAD

**Issue by rprechelt** [JuliaLang/METADATA.jl#5128]
Over the last year, the clone time for METADATA.jl has grown significantly, (it is now in excess of a minute for me on a reasonably fast connection) and using the Github page is a total nightmare due to the slow response time; I'm sure that I'm not the only person who has experienced this. 

Obviously, this problem is only going to get worse as the number of packages in METADATA grows and Julia becomes more popular (and hence more requests to METADATA). I would hate for us to end up like CocoaPods - being [rate-limited](https://github.com/CocoaPods/CocoaPods/issues/4989) by Github 24/7 and taking 20+ minutes to clone. Obviously, the structure and operation of METADATA as compared to CocoaPods is different, but some of the comments and discussions by Github infrastructure engineers with the CocoaPods team about increasing performance might be somewhat relevant to us. 

In particular, this [comment](https://github.com/CocoaPods/CocoaPods/issues/4989#issuecomment-193835710), by a Github engineer, strongly recommends sharding the root repository by the first two letters to drastically improve clone time and page access, while reducing the load on Github's servers. 

I am not intimately familiar with how the Julia Package Manager interacts with METADATA, but it seems like any change like this might best be done when Julia is still relatively young and before what is currently an issue becomes a problem. We are also coming up to a new release (v0.5), and might be able to sneak this in then. 

Comments? Thoughts? Ideas?

**Comment by johnmyleswhite** [comment]
I tend to think the package system should not be using a version control system for this task, but instead a database. While it's tempting to be able to refer to historical versions of METADATA, I suspect the history that you usually want is present in the current state of the repo, rather than in its historical state.

**Comment by tkelman** [comment]
There is value in having the history of METADATA available - if we kept a local activity log of which past versions are used on `Pkg.update()` then you'd be able to roll back and at least partially "undo" an update by checking out an older state of METADATA and re-running the resolver.

We should eventually move to a modified system for this, and it'll be almost impossible to do so in a backwards compatible way unless we eventually lock the "old" repository/branch to all changes that don't come from an automatic back-compat mirroring bot. Such a new system needs to be designed and implemented first, then maintained on an ongoing basis.

Github-based systems seem to work moderately well for homebrew and Rust's crates (https://github.com/rust-lang/crates.io-index), which are both several times larger than METADATA is in number of packages. Though homebrew overwrites old versions of packages, and Rust's crates system looks to have been designed a bit more carefully - one file per package that contains all information about all versions, sharded directory structure, etc.

**Comment by rprechelt** [comment]
So the current plan is to hold off any improvements (such as sharding) until a more complete and sophisticated system can be designed?

**Comment by tkelman** [comment]
I certainly think it's far too late to do anything before 0.5. A major Pkg redesign is planned, hopefully for 0.6, and this should be one piece of it.

**Comment by tkelman** [comment]
It won't help with initial clone time, but one thing that we could do immediately that would be totally non-disruptive and help save network traffic and git cpu cycles would be switching the "has upstream METADATA been updated" check to use the dedicated Github API that exists for this purpose: https://developer.github.com/changes/2016-02-24-commit-reference-sha-api

**Comment by rprechelt** [comment]
That sounds within the realm of 0.5 possibilities; if I get the time this week, I'll see if I can't crank out a PR for it.

**Comment by TotalVerb** [comment]
How many clones of METADATA are from CI? If that's a substantial amount, would it help to suggest to use caching (https://github.com/travis-ci/travis-ci/issues/5744) on Travis for METADATA.jl, so that the repository only needs to be updated, and not cloned?

**Comment by tkelman** [comment]
I suppose you could cache the metadata directory, but you'll want to add a call to `Pkg.update()` if you do so to make sure you get the latest versions of your dependencies. And be aware that branches and pull requests upload new cache versions if any cached content changes, so you may check the travis cache settings a few weeks later for a busy package and see dozens of copies for no-longer-relevant branches or PR's that you'd have to clear manually.

**Comment by TotalVerb** [comment]
Is clone time becoming slower for anyone else?

**Comment by KristofferC** [comment]
FWIW I have a "sharded" METADATA at https://github.com/KristofferC/METADATA.jl and a Pkg replacement that works with that metadata https://github.com/KristofferC/Pkg25.jl. I originally made it for https://github.com/JuliaLang/julia/pull/18405 but now pkg3 is on its way I guess so I didn't move forward on it. Not really advocating anyone to use it but maybe it can be used as a timing reference.

**Comment by tkelman** [comment]
Pkg25Dev doesn't exist, right? So would have to switch back to metadata-v2 to do any PkgDev operations? Luckily attobot makes doing those locally unnecessary, but adoption of attobot isn't as close to 100% yet as it should be.

**Comment by KristofferC** [comment]
No it doesn't exist and I am not really planning on creating it either. I think that something like attobot will make a lot of the stuff PkgDev obsolete (except generating the package file skeleton).

**Comment by mehmetaergun** [comment]
Thanks for all your work! 

Just as a side note, having waited for about 12 minutes for the initial clone, you might want to also consider the impact of anti-virus software trying to block each file being downloaded, probably exponentially increasing the wait time. I see about 27,000 files taking up 130MB in the config directory (~/.julia). 

This is a Julia 0.6.1 first time installation. 

Btw, what happens when github goes out of business etc?

**Comment by KristofferC** [comment]
The METADATA repo can be hosted anywhere. The replacement of METADATA for Julia 1.0 will have much fewer files and greatly improve this.

**Comment by constantza** [comment]
Hi,
I'm trying to install the last stable version (0.6.2) of Julia. 
I'm in Windows 7 x64, I have a 4Ghz i5 4690k, 32 Go of DDR3 1600 Mhz CL8, so not old config. I'm able to download 1Go from Youtube in 5 minutes.

The cloning of the METADATA.jl seems to be excessively long. I deactivated the firewall and the antivirus (is dangerous to connect to go to other websites in parallel without if I must rest more than 10-20 minutes). 

The process started 25 minutes ago, and is Julia seems to be blocked. For sure not, but however, when I connected to https://github.com/JuliaLang/METADATA.jl, I have seen a huuuuuuge list of files to clone. Even when I asked Firefox to clone physically the ZIP, the 6Mo take 5 minutes. And when I tried to unzip it, something was wrong because of error in headers. However I may unzip (I don't know if correctly, this is another problem, out of the discussion here) and detected 26485 files with a real volum more than 100 MB).

Finally the process of cloning stopped (in all was 32 min), but unfortunately with an error message:

INFO: Initializing package repository C:\Users\User\.julia\v0.6
INFO: Cloning METADATA from https://github.com/JuliaLang/METADATA.jl
INFO: Cloning cache of ArrayFire from https://github.com/JuliaComputing/ArrayFire.jl.git
INFO: Cloning cache of BinDeps from https://github.com/JuliaLang/BinDeps.jl.git
ERROR: Cannot clone BinDeps from https://github.com/JuliaLang/BinDeps.jl.git. Failed to receive response: Le serveur a renvoyé une réponse non valide ou non reconnue

(sorry, I'm in France, so the last message was in french, for informing that the server sent a non valid or not recognized answer).

We are in 2018, Julia is not yet 1.0 (just 0.6.2) but is annoying to add all the problems for ending with a normal environment. Wanting to use a quick environment as Julia but adding days or weeks (or even months if we take in consideration the systematic incompatibilities between the packages and the versions of Julia) is not very attractive. 

I respect the work of all people evolved in the project. But I must admit that for R in 10 years a never had to fight this way for installing the software or the packages from CRAN repository. Something must change because is sad for some environment of the quality of Julia to be penalized this way.

Best regards

**Comment by KristofferC** [comment]
Yes, this is changing for the 1.0 release. The new METADATA will look like https://github.com/JuliaRegistries/Uncurated.

It has significantly fewer files and the new package manager (https://github.com/JuliaLang/Pkg3.jl) is implemented in a way that reads fewer files. For a speed comparison between the current package manager and the new one, see e.g:

https://www.youtube.com/watch?v=s8iRe_pSVYE

**Comment by DilumAluthge** [comment]
METADATA.jl is no longer maintained, so I am closing this issue.

If this issue is still relevant, please re-open a new issue on the Julia General registry: https://github.com/JuliaRegistries/General
